queasars/job_shop_scheduling/scip_solver.py

```diff
@@ -0,0 +1,136 @@
+                # This is needed for the additional constraints
+                if operation.machine not in self._machine_operations:
+                    self._machine_operations[operation.machine] = []
+                self._machine_operations[operation.machine].append(operation)
+
+                # Create an integer variable for the start time of each operation
+                # Constrain it to be at least 0
+                var = self._model.addVar(operation.identifier, vtype="INTEGER")
+                self._operation_start_variables[operation] = var
+                self._model.addCons(var >= 0)
+
+    def _prepare_constraints(self):
+        """
+        Constrains all operations within each job to start in order and not overlap.
+        Constrains all operations for each machine to not overlap. To this end auxiliary binary variables
+        are created to enable ordering decisions.
+        """
+        for job in self._jssp_instance.jobs:
+            n_operations = len(job.operations)
+
+            # Ensure by constraint, that each operation in a job finishes, before the next one may start
+            for i in range(0, n_operations - 1):
+                self._model.addCons(
+                    self._operation_start_variables[job.operations[i]] + job.operations[i].processing_duration
+                    <= self._operation_start_variables[job.operations[i + 1]]
+                )
+            # For each job the makespan must be larger than the end time of its last operation
+            self._model.addCons(
+                self._operation_start_variables[job.operations[n_operations - 1]]
+                + job.operations[n_operations - 1].processing_duration
+                <= self._optimization_var
+            )
+
+        # Ensure for each machine, that no operation may overlap
+        for operations in self._machine_operations.values():
+
+            # For a machine look at all possible combination of two operations
+            for operation_1, operation_2 in combinations(operations, 2):
+                # Introduce a binary variable to allow the choice of precedence between the two operations
+                order_var = self._model.addVar(
+                    f"order_{operation_1.identifier}_{operation_2.identifier}", vtype="BINARY"
+                )
+                # Depending on the precedence order, ensure that the later operation may only start, once the earlier
+                # one has finished
+                self._model.addCons(
+                    order_var * (self._operation_start_variables[operation_1] + operation_1.processing_duration)
+                    <= self._operation_start_variables[operation_2]
+                )
+                self._model.addCons(
+                    (1 - order_var) * (self._operation_start_variables[operation_2] + operation_2.processing_duration)
+                    <= self._operation_start_variables[operation_1]
+                )
```

Comment:
Nice solution! It's possible, however, that SCIP offers something for this as there are `Model.addConsOr` as well as `Model.addConsXor`. I don't have any experience with them but maybe you can have a quick look if this may simplify (or accelerate) the SCIP solution?

Reply:
From how i understand it, addConsOr is an or constraint on a bunch of boolean variables. This would need the introduction of auxiliary boolean variables as well. There is apparently a way to connect constraints using a disjunction in SCIP [(scip-documentation),](https://scipopt.org/doc/html/cons__disjunction_8c.php) but this functionality does not seem to be implemented in pyscipopt. As such, i would suggest to leave this part of the scip solver as is for now.

Reply:
Agree, thanks for looking this up, though!